Basic_Python_Diploma/VK.py

```diff
@@ -165,49 +141,31 @@ def mutual_friends(self, friend):
         mutual_friends_params.update(self.params)
         ids_list = requests.get(self.API_URL + self.methods['friends']['getMutual'],
                                 params=mutual_friends_params).json()['response']
-
-        ids_str = ''
-        friends_list = [User(_id) for _id in ids_list]
-        for _id in ids_list:
-            if not ids_str:
-                ids_str += str(_id)
-            else:
-                ids_str = ids_str + "," + str(_id)
-        mut_friends_names = self.user(ids_str)
-        mut_friends_names_list = []
-        for _name in mut_friends_names:
-            name = _name['first_name'] + ' ' + _name['last_name']
-            mut_friends_names_list.append(name)
-        print(f'{self.user()} и {User(friend).user()} имеют {len(mut_friends_names_list)} общих друзей:')
-        print(*mut_friends_names_list, sep=", ")
+        time.sleep(0.5)
+        friends_list = tqdm([User(_id).name for _id in ids_list], desc="Получение списка общих друзей.")
```

Comment:
Этот прогресс-бар будет врать. Сначала полностью выпонится list comprehension и будет составлен список -- это по сути и есть вся медленная работа, которую и стоило бы показывать в прогресс-баре. А затем прогресс-бар моментально дойдёт до 100 процентов.

Думаю, правильнее было бы сделать вот так:

```suggestion
        friends_list = tqdm((User(_id).name for _id in ids_list), total=len(ids_list), desc="Получение имён общих друзей.")
```

(Код не проверял, так что возможны конфузы).

Выражение в круглых скобках -- это уже не list comprehension, а генераторное выражение. Разница в том, что list comprehension выполняется энергично (eager) до конца, сразу же как только его определили. Генераторное выражение выполняется лениво -- оно начинает работать только тогда, когда по нему кто-то начнёт итерироваться. Но так как у генераторного выражения нельзя заранее узнать длину, то приходится в `tqdm` явно передавать количество элементов, чтобы он смог правильно отобразить прогресс-бар.

На почитать:
* https://docs.python.org/3/howto/functional.html#generator-expressions-and-list-comprehensions
* https://stackoverflow.com/questions/47789/generator-expressions-vs-list-comprehension